src/lib/Inference.hs

```diff
@@ -736,16 +736,17 @@ checkOrInferRho (WithSrcE pos expr) reqTy = do
           UPatBinder UIgnore ->
             buildPiInf noHint arr ann' \_ -> (,) <$> checkUEffRow effs <*> checkUType ty
           _ -> buildPiInf (getNameHint pat) arr ann' \v -> do
-            Abs decls piResult <- buildDeclsInf do
+            Abs decls (effs'' `PairE` result) <- buildDeclsInf do
               v' <- sinkM v
               bindLamPat (WithSrcB pos' pat) v' do
                 effs' <- checkUEffRow effs
                 ty'   <- checkUType   ty
                 return $ PairE effs' ty'
-            cheapReduceWithDecls decls piResult >>= \case
+            cheapReduceWithDecls decls (effs'' `PairE` result) >>= \case
               (Just (PairE effs' ty'), DictTypeHoistSuccess, []) -> return $ (effs', ty')
               _ -> throw TypeErr $ "Can't reduce type expression: " ++
-                     pprint (Block (BlockAnn TyKind) decls $ snd $ fromPairE piResult)
+                     -- Coercing for printing only
+                     pprint (Block (BlockAnn TyKind $ unsafeCoerceE effs'') decls result)
```

Comment:
Well, status quo is that the Pretty instance of Block does not, in fact, read the annotation.  Encoded that in a separate `prettyBlock` function which can be used in a place like this to avoid constructing a bogus annotation.